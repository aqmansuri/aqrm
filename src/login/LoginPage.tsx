import React from 'react';
import {
  IonButton,
  IonContent,
  IonFooter,
  IonHeader,
  IonIcon,
  IonInput,
  IonItem,
  IonLabel,
  IonList,
  IonPage,
  IonTitle,
  IonToolbar,
} from '@ionic/react';
import { useForm, Controller } from 'react-hook-form';
import { logInOutline } from 'ionicons/icons';

const LoginPage: React.FC = () => {
  const {
    handleSubmit,
    control,
    formState: { errors, isDirty, isValid },
  } = useForm({ mode: 'onTouched' });

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>Login</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent>
        <IonHeader collapse="condense">
          <IonToolbar>
            <IonTitle size="large">Login</IonTitle>
          </IonToolbar>
        </IonHeader>
        <form>
          <IonList>
            <IonItem>
              <IonLabel position="floating">E-Mail Address</IonLabel>
              <Controller
                render={({ field: { onChange, value, onBlur } }) => (
                  <IonInput
                    data-testid="email-input"
                    onIonChange={e => onChange(e.detail.value!)}
                    onIonBlur={onBlur}
                    value={value}
                    type="email"
                  />
                )}
                control={control}
                name="email"
                rules={{
                  required: true,
                  pattern: {
                    value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
                    message: 'E-Mail Address must have a valid format',
                  },
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel position="floating">Password</IonLabel>
              <Controller
                render={({ field: { onChange, value, onBlur } }) => (
                  <IonInput
                    data-testid="password-input"
                    onIonChange={e => onChange(e.detail.value!)}
                    onIonBlur={onBlur}
                    value={value}
                    type="password"
                  />
                )}
                control={control}
                name="password"
                rules={{ required: true }}
              />
            </IonItem>
          </IonList>
          <div className="error-message" data-testid="errors">
            <div>
              {errors.email?.type === 'required' &&
                'E-Mail Address is required'}
            </div>
            <div>
              {errors.email?.type === 'pattern' && errors.email.message}
            </div>
            <div>
              {errors.password?.type === 'required' && 'Password is required'}
            </div>
          </div>
        </form>
      </IonContent>
      <IonFooter>
        <IonToolbar>
          <IonButton
            expand="full"
            disabled={!isDirty || !isValid}
            onClick={handleSubmit(data => console.log(data))}
            data-testid="submit-button"
          >
            Sign In
            <IonIcon slot="end" icon={logInOutline} />
          </IonButton>
        </IonToolbar>
      </IonFooter>
    </IonPage>
  );
};

export default LoginPage;
