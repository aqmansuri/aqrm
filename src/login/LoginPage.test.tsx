import { render, waitFor } from '@testing-library/react';
import { ionFireEvent as fireEvent } from '@ionic/react-test-utils';
import LoginPage from './LoginPage';

describe('<LoginPage />', () => {
  it('displays the header', () => {
    const { container } = render(<LoginPage />);
    expect(container).toHaveTextContent(/Login/);
  });

  it('renders consistently', () => {
    const { asFragment } = render(<LoginPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('sign in button', () => {
    it('starts disabled', () => {
      const { getByTestId } = render(<LoginPage />);
      const button = getByTestId(/submit-button/) as HTMLIonButtonElement;
      expect(button.disabled).toBeTruthy();
    });

    it('is disabled with just an e-mail address', async () => {
      const { getByTestId } = render(<LoginPage />);
      const button = getByTestId(/submit-button/) as HTMLIonButtonElement;
      const email = getByTestId(/email-input/);
      await waitFor(() => {
        fireEvent.ionChange(email, 'test@test.com');
      });
      expect(button.disabled).toBeTruthy();
    });

    it('is disabled with just a password', async () => {
      const { getByTestId } = render(<LoginPage />);
      const button = getByTestId(/submit-button/) as HTMLIonButtonElement;
      const password = getByTestId(/password-input/);
      await waitFor(() => {
        fireEvent.ionChange(password, 'P@ssword123');
      });
      expect(button.disabled).toBeTruthy();
    });

    it('is enabled with both an email address and a password', async () => {
      const { getByTestId } = render(<LoginPage />);
      const button = getByTestId(/submit-button/) as HTMLIonButtonElement;
      const email = getByTestId(/email-input/);
      const password = getByTestId(/password-input/);
      await waitFor(() => {
        fireEvent.ionChange(email, 'test@test.com');
        fireEvent.ionChange(password, 'P@ssword123');
      });
      expect(button.disabled).toBeFalsy();
    });
  });

  describe('error messages', () => {
    it('starts with no error messages', () => {
      const { getByTestId } = render(<LoginPage />);
      const errors = getByTestId(/errors/);
      expect(errors).toHaveTextContent('');
    });

    it('displays an error if the e-mail address is dirty and empty', async () => {
      const { getByTestId } = render(<LoginPage />);
      const errors = getByTestId(/errors/);
      const email = getByTestId(/email-input/);
      await waitFor(() => {
        fireEvent.ionChange(email, 'test@test.com');
        fireEvent.ionChange(email, '');
      });
      expect(errors).toHaveTextContent(/E-Mail Address is required/);
    });

    it('displays an error message if the e-mail address has an invalid format', async () => {
      const { getByTestId } = render(<LoginPage />);
      const errors = getByTestId(/errors/);
      const email = getByTestId(/email-input/);
      await waitFor(() => {
        fireEvent.ionChange(email, 'foo');
      });
      expect(errors).toHaveTextContent(
        /E-Mail Address must have a valid format/,
      );
    });

    it('displays an error message if the password is dirty and empty', async () => {
      const { getByTestId } = render(<LoginPage />);
      const errors = getByTestId(/errors/);
      const password = getByTestId(/password-input/);
      await waitFor(() => {
        fireEvent.ionChange(password, 'P@ssword123');
        fireEvent.ionChange(password, '');
      });
      expect(errors).toHaveTextContent(/Password is required/);
    });
  });
});
