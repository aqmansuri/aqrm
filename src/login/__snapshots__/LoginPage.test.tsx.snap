// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-header>
      <ion-toolbar>
        <ion-title>
          Login
        </ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content>
      <ion-header
        collapse="condense"
      >
        <ion-toolbar>
          <ion-title
            size="large"
          >
            Login
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <form>
        <ion-list>
          <ion-item>
            <ion-label
              position="floating"
            >
              E-Mail Address
            </ion-label>
            <ion-input
              data-testid="email-input"
              type="email"
            />
          </ion-item>
          <ion-item>
            <ion-label
              position="floating"
            >
              Password
            </ion-label>
            <ion-input
              data-testid="password-input"
              type="password"
            />
          </ion-item>
        </ion-list>
        <div
          class="error-message"
          data-testid="errors"
        >
          <div />
          <div />
          <div />
        </div>
      </form>
    </ion-content>
    <ion-footer>
      <ion-toolbar>
        <ion-button
          data-testid="submit-button"
          disabled="true"
          expand="full"
        >
          Sign In
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Log In</title><path d='M192 176v-40a40 40 0 0140-40h160a40 40 0 0140 40v240a40 40 0 01-40 40H240c-22.09 0-48-17.91-48-40v-40' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><path stroke-linecap='round' stroke-linejoin='round' d='M288 336l80-80-80-80M80 256h272' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            slot="end"
          />
        </ion-button>
      </ion-toolbar>
    </ion-footer>
  </div>
</DocumentFragment>
`;
